xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractror.java

```diff
@@ -0,0 +1,476 @@
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+import org.apache.parquet.LogicalType;
+import org.apache.parquet.LogicalTypes;
+//import org.apache.parquet.Schema;
+import org.apache.parquet.Schema.Type;
+import org.apache.parquet.schema.PrimitiveType;
+import org.apache.parquet.format.NullType;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
+
+
+import org.apache.xtable.collectors.CustomCollectors;
+import org.apache.xtable.exception.SchemaExtractorException;
+import org.apache.xtable.exception.UnsupportedSchemaTypeException;
+import org.apache.xtable.hudi.idtracking.IdTracker;
+import org.apache.xtable.hudi.idtracking.models.IdMapping;
+import org.apache.xtable.model.schema.InternalField;
+import org.apache.xtable.model.schema.InternalSchema;
+import org.apache.xtable.model.schema.InternalType;
+import org.apache.xtable.schema.SchemaUtils;
+
+import org.apache.xtable.avro.AvroSchemaConverter;
+import org.apache.avro.Schema;
+//import org.apache.parquet.avro.AvroSchemaConverter;
+
+/**
+ * Class that converts parquet Schema {@link Schema} to Canonical Schema {@link InternalSchema} and
+ * vice-versa. This conversion is fully reversible and there is a strict 1 to 1 mapping between
+ * parquet data types and canonical data types.
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class ParquetSchemaExtractor {
+    // parquet only supports string keys in maps
+    private static final InternalField MAP_KEY_FIELD =
+            InternalField.builder()
+                    .name(InternalField.Constants.MAP_KEY_FIELD_NAME)
+                    .schema(
+                            InternalSchema.builder()
+                                    .name("map_key")
+                                    .dataType(InternalType.STRING)
+                                    .isNullable(false)
+                                    .build())
+                    .defaultValue("")
+                    .build();
+    private static final ParquetSchemaExtractor INSTANCE = new ParquetSchemaExtractor();
+    private static final String ELEMENT = "element";
+    private static final String KEY = "key";
+    private static final String VALUE = "value";
+
+    public static ParquetSchemaExtractor getInstance() {
+        return INSTANCE;
+    }
+
+    private static Type finalizeSchema(MessageType targetSchema, InternalSchema inputSchema) {
+        if (inputSchema.isNullable()) {
+            return targetSchema.union(LogicalTypeAnnotation.unknownType())
+        }
+        return targetSchema;
+    }
+
+    private static boolean groupTypeContainsNull(Type schema) {
+        for (Type field in schema.getFields()){
+            if (field == null) {
+                return True;
+            }
+        }
+        return False;
+    }
+
+    public InternalSchema _toInternalSchema(Schema schema) {
```

Comment:
The `_` can be removed from this method name to follow traditional java naming conventions